File: ios/Tehilim/Core/Services/HebrewDateFormatter.swift
import Foundation

enum HebrewDateFormatter {
    struct DisplayDate {
        let dayOfWeek: String          // "Lundi" / "Monday"
        let transliterated: String     // "7 Iyar 5786"
        let hebrew: String             // "ז׳ באייר ה׳תשפ״ו"
    }

    /// Résout la locale du jour de la semaine depuis `pref.app.language` —
    /// indispensable pour suivre la bascule de langue à chaud (V2.1.b). Avant,
    /// `Locale(identifier: "fr_FR")` était hardcodé, donc « Lundi » restait
    /// FR même quand l'app passait en EN.
    private static func dayOfWeekLocale() -> Locale {
        let raw = AppGroup.userDefaults.string(forKey: AppGroup.Keys.appLanguage)
            ?? UserDefaults.standard.string(forKey: "pref.app.language")
            ?? "system"
        switch raw {
        case "fr": return Locale(identifier: "fr_FR")
        case "en": return Locale(identifier: "en_US")
        case "he": return Locale(identifier: "he_IL")
        default:
            // .system : on lit `Locale.preferredLanguages` (live) et NON
            // `Locale.current` (instantané figé qui reste sur l'ancienne langue
            // après une bascule à chaud → jour de la semaine en anglais alors
            // que le système est en français). Hébreu / français si l'appareil
            // est dans ces langues, anglais sinon.
            let pref = Locale.preferredLanguages.first ?? "fr"
            if pref.hasPrefix("he") || pref.hasPrefix("iw") { return Locale(identifier: "he_IL") }
            return pref.hasPrefix("fr") ? Locale(identifier: "fr_FR") : Locale(identifier: "en_US")
        }
    }

    static func formatted(_ date: Date = Date()) -> DisplayDate {
        let dowFmt = DateFormatter()
        dowFmt.locale = dayOfWeekLocale()
        dowFmt.dateFormat = "EEEE"
        let day = dowFmt.string(from: date).capitalized(with: dowFmt.locale)

        let latinFmt = DateFormatter()
        latinFmt.calendar = Calendar(identifier: .hebrew)
        latinFmt.locale = Locale(identifier: "en_US_POSIX")
        latinFmt.dateFormat = "d MMMM yyyy"
        let latin = latinFmt.string(from: date)

        let heFmt = DateFormatter()
        heFmt.calendar = Calendar(identifier: .hebrew)
        heFmt.locale = Locale(identifier: "he_IL")
        heFmt.dateStyle = .long
        let hebrew = heFmt.string(from: date)

        return DisplayDate(dayOfWeek: day, transliterated: latin, hebrew: hebrew)
    }
}

File: android/app/src/main/java/com/david/tehilim/core/service/HebrewDateFormatter.kt
package com.david.tehilim.core.service

import android.icu.text.DateFormat
import android.icu.text.SimpleDateFormat
import android.icu.util.Calendar
import android.icu.util.HebrewCalendar
import android.icu.util.ULocale
import androidx.appcompat.app.AppCompatDelegate
import java.util.Date
import java.util.Locale

/**
 * Mirror du HebrewDateFormatter iOS.
 *
 * Utilise **android.icu** (Unicode CLDR) qui supporte officiellement le
 * calendrier hébraïque et son formatage en hébreu (כ״ז באייר ה׳תשפ״ו)
 * et en translittération latine (27 Iyar 5786).
 *
 * Disponible depuis API 24, donc compatible avec notre min SDK 26.
 */
object HebrewDateFormatter {

    data class DisplayDate(
        val dayOfWeek: String,         // « Jeudi » ou « Thursday »
        val transliterated: String,    // « 27 Iyar 5786 »
        val hebrew: String             // « כ״ז באייר ה׳תשפ״ו »
    )

    /**
     * Résout la locale courante de l'app (FR/EN) à partir d'AppCompatDelegate.
     * V1.3.8 — fixe le bug « Mercredi » en anglais : avant, on hardcodait
     * `Locale.FRENCH` au lieu de lire la locale active.
     */
    private fun currentLocale(): Locale {
        // Sur API 33+ l'app pilote la langue via LocaleManager, donc
        // getApplicationLocales() peut être vide → on retombe sur Locale.getDefault()
        // (qui reflète bien la langue de l'app). On normalise ensuite.
        val locales = AppCompatDelegate.getApplicationLocales()
        val base = if (!locales.isEmpty) (locales[0] ?: Locale.getDefault()) else Locale.getDefault()
        // `Locale.getLanguage()` rapporte l'hébreu sous l'ancien code « iw ».
        return when (base.language) {
            "en" -> Locale.ENGLISH
            "fr" -> Locale.FRENCH
            "he", "iw" -> Locale("he")
            else -> Locale.FRENCH
        }
    }

    fun formatted(date: Date = Date()): DisplayDate {
        val locale = currentLocale()
        // Jour de la semaine dans la locale courante (« lundi » ou « Monday »)
        val dowFmt = SimpleDateFormat("EEEE", locale)
        val day = dowFmt.format(date).replaceFirstChar { it.titlecase(locale) }

        // Date hébraïque translittérée (anglais — « 27 Iyar 5786 »)
        val latinCal = HebrewCalendar()
        latinCal.time = date
        val latinFmt = DateFormat.getDateInstance(DateFormat.LONG, ULocale("en_US@calendar=hebrew"))
        latinFmt.calendar = latinCal
        val latin = latinFmt.format(date)

        // Date hébraïque en hébreu (« כ״ז באייר ה׳תשפ״ו »)
        val heCal = HebrewCalendar()
        heCal.time = date
        val heFmt = DateFormat.getDateInstance(DateFormat.LONG, ULocale("he_IL@calendar=hebrew"))
        heFmt.calendar = heCal
        val hebrew = heFmt.format(date)

        return DisplayDate(dayOfWeek = day, transliterated = latin, hebrew = hebrew)
    }

    /**
     * Pour le widget : version compacte « 27 Iyar » sans année.
     */
    fun compact(date: Date = Date()): String {
        val cal = HebrewCalendar()
        cal.time = date
        val fmt = SimpleDateFormat("d MMMM", Locale.ENGLISH)
        fmt.calendar = cal
        return fmt.format(date)
    }
}
